### ISSUE TITLE
训练效率问题

### ISSUE THREAD

**Issue by ScottishFold007** [Morizeyao/GPT2-Chinese#130]
我用笔记本电脑的GTX1070（显存8G），训练了一个虎嗅新闻语料，约110MB，也能跑，每一步保存模型，时间间隔是5分钟。想看看大家的机器及训练效率，欢迎在下方留言探讨。我的参数设置如下：
{
  "architectures": [
    "GPT2LMHeadModel"
  ],
  "attn_pdrop": 0.1,
  "bos_token_id": 0,
  "do_sample": false,
  "embd_pdrop": 0.1,
  "eos_token_ids": 0,
  "finetuning_task": null,
  "id2label": {
    "0": "LABEL_0"
  },
  "initializer_range": 0.02,
  "is_decoder": false,
  "label2id": {
    "LABEL_0": 0
  },
  "layer_norm_epsilon": 1e-05,
  "length_penalty": 1.0,
  "max_length": 20,
  "model_type": "gpt2",
  "n_ctx": 1024,
  "n_embd": 768,
  "n_head": 6,
  "n_layer": 12,
  "n_positions": 1024,
  "num_beams": 1,
  "num_labels": 1,
  "num_return_sequences": 1,
  "output_attentions": false,
  "output_hidden_states": false,
  "output_past": true,
  "pad_token_id": 0,
  "pruned_heads": {},
  "repetition_penalty": 1.0,
  "resid_pdrop": 0.1,
  "summary_activation": null,
  "summary_first_dropout": 0.1,
  "summary_proj_to_labels": true,
  "summary_type": "cls_index",
  "summary_use_proj": true,
  "temperature": 1.0,
  "top_k": 50,
  "top_p": 1.0,
  "torchscript": false,
  "use_bfloat16": false,
  "vocab_size": 21029
}

其他参数设置：
 _adam_epsilon=1e-08, block_size=500, cache_dir=None, config_name='C:\\Users\\hp\\Desktop\\2020.02.15 GPT2模型从头开始训练modeloutput\\GPT2LMHeadModel\\config.json', device=device(type='cuda'), do_eval=False, do_train=True, eval_all_checkpoints=False, eval_data_file=None, evaluate_during_training=False, fp16=False, fp16_opt_level='O1', gradient_accumulation_steps=1, learning_rate=5e-05, line_by_line=True, local_rank=-1, logging_steps=500, max_grad_norm=1.0, max_steps=-1, mlm=False, mlm_probability=0.15, model_name_or_path=None, model_type='gpt2', n_gpu=1, no_cuda=False, num_train_epochs=40.0, output_dir='D:\\20200215OUTPUT', overwrite_cache=False, overwrite_output_dir=True, per_gpu_eval_batch_size=4, per_gpu_train_batch_size=4, save_steps=500, save_total_limit=10, seed=42, server_ip='', server_port='', should_continue=False, tokenizer_name=None, train_data_file='huxiu.txt', warmup_steps=0, weight_decay=0.0_
跑40轮，目前的生成效果：

**个人认为其实核心不是一个场景代替另一个场景，而是用户的行为有没有被数字化。**
这种观点可能会让我们先来看下它对于未知之甚少大脑中所提到：1、社交与移动互联网连接在相当长时间内形成强关系（比如微博）；2、信息通讯录方式决定商业价值变现模型需要重新设计产品和服务体验；3、随着智慧版本发布以及更多地区域限制改善了图片分享功耗电子邮件等优势；4.基础架构也就意味深层出不穷或者专门针对性别问题进入该领导原因 ...

**个人认为o2o**。如果你是一家公司，那么可能就不要成功了；或者说它的模式也许只有两种：第三方服务商、支付与互联网基础设施；第四类型应用来自于线上和线下结合。这些都没什么区别二维码（大数据分析），但其实在我看来还是很清楚地解决这样问题。首先从本质出发点而言，目前大多数中小企业正处于相关领域之间建立起强烈竞争对手的位置，并且无法获得更好的收益率以及他们所需要做到的事情，因此，这里面提供了最简单粗暴力量的 ...

**Comment by angletxm** [comment]
我用gpt2训练对联，跑了一千轮了要 还是没有达到拟合，请问是训练的不够吗 ？
def __init__(
        self,
        vocab_size=2341,
        n_positions=128,
        n_ctx=128,
        n_embd=128,
        n_layer=4,
        n_head=8,
        resid_pdrop=0.1,
        embd_pdrop=0.1,
        attn_pdrop=0.1,
        layer_norm_epsilon=1e-5,
        initializer_range=0.02,
        output_hidden_states=False,
        output_attentions=False,
        lr=3e-5,
        dynamics_lr=False,
        read_len=40000,
        history_len=5,
        **kwargs
    )

**Comment by ScottishFold007** [comment]
模型缩水了。。。，参数设置有问题



---原始邮件---
发件人: "angletxm"<notifications@github.com&gt;
发送时间: 2020年10月30日(周五) 下午5:24
收件人: "Morizeyao/GPT2-Chinese"<GPT2-Chinese@noreply.github.com&gt;;
抄送: "Author"<author@noreply.github.com&gt;;"Scottish_Fold007"<gaochangkuan01@foxmail.com&gt;;
主题: Re: [Morizeyao/GPT2-Chinese] 训练效率问题 (#130)




 
我用gpt2训练对联，跑了一千轮了要 还是没有达到拟合，请问是训练的不够吗 ？
 def init(
 self,
 vocab_size=2341,
 n_positions=128,
 n_ctx=128,
 n_embd=128,
 n_layer=4,
 n_head=8,
 resid_pdrop=0.1,
 embd_pdrop=0.1,
 attn_pdrop=0.1,
 layer_norm_epsilon=1e-5,
 initializer_range=0.02,
 output_hidden_states=False,
 output_attentions=False,
 lr=3e-5,
 dynamics_lr=False,
 read_len=40000,
 history_len=5,
 **kwargs
 )

**Comment by BiEchi** [comment]
我单机四块V100,用的DistributedDataParallel+Apex,10秒一个epoch,5m的语料库(0.1B)一小时基本拟合完成.
